core/src/chain/memory/mod.rs

```diff
@@ -0,0 +1,368 @@
+use std;
+
+use chain::entry::Entry;
+use chain::pair::Pair;
+use chain::SourceChain;
+
+#[derive(Clone, Debug, PartialEq, Default, Serialize, Deserialize)]
+pub struct MemChain {
+    pairs: Vec<Pair>,
+    top: Option<Pair>,
+}
+
+impl MemChain {
+    /// builds the data structures required to efficiently represent a SourceChain in memory
+    /// typically a MemChain should be _mutable_ to facilitate chain.push()
+    pub fn new() -> MemChain {
+        MemChain {
+            pairs: Vec::new(),
+            top: None,
+        }
+    }
+}
+
+/// SouceChain trait implementation
+impl<'de> SourceChain<'de> for MemChain {
+    // appends the current pair to the top of the chain
+    fn push(&mut self, entry_type: &str, entry: &Entry) -> Pair {
+        let pair = Pair::new(self, entry_type, entry);
+
+        if !(pair.validate()) {
+            panic!("attempted to push an invalid pair for this source chain");
```

Comment:
yar, i know, but atm it can only happen at dev time

there's a ticket up to reassess this once zome validation is a thing - https://github.com/holochain/holochain-rust/issues/63